frame/assets/src/lib.rs

```diff
@@ -210,14 +220,16 @@ pub mod pallet {
 		type RemoveItemsLimit: Get<u32>;
 
 		/// Identifier for the class of asset.
-		type AssetId: Member
-			+ Parameter
-			+ Default
+		type AssetId: Member + Parameter + Copy + MaybeSerializeDeserialize + MaxEncodedLen;
+
+		/// Wrapper around `Self::AssetId` to use in dispatchable call signatures. Allows the use
+		/// of compact encoding in instances of the pallet, which will prevent breaking changes
+		/// resulting from the removal of `HasCompact` from `Self::AssetId`.
+		type AssetIdParameter: Parameter
 			+ Copy
-			+ HasCompact
-			+ MaybeSerializeDeserialize
-			+ MaxEncodedLen
-			+ TypeInfo;
+			+ From<Self::AssetId>
```

Comment:
Can we remove the `From<Self::AssetId>` bound and only keep `Into<Self::AssetId>` here?

Reply:
It is possible, made a branch here: https://github.com/paritytech/substrate/compare/joe-assetid-bounds...joe-assetid-bounds-no-from?expand=1

Let me know if this is better and I will merge it into this.

Reply:
I added a commit ([46dfc28](https://github.com/paritytech/substrate/pull/12740/commits/46dfc286808fefa5bb3f991a5cd3404fd54087e1)) to do this, but I'm not sure if it's the right idea. On one hand, it makes sense to go only from the param to everything else, without the need to go "backwards".

But some runtimes might have pallets tightly coupled with Assets, and call these functions directly. In those cases, they'd want to be able to go from an ID to a param.

Can you explain more why you think we should do this?

Reply:
Including `From` for this PR as it's the safer (although more restrictive) configuration for coupling.

Reply:
> Can you explain more why you think we should do this?

At first, I thought that `From<Self::AssetId>` and `Into<Self::AssetId>` were duplicates and thought that it would be enough to just keep `From<Self::AssetId>`. However, after looking through the documentation, I think you're right. they are two different things.

According to [From docs](https://doc.rust-lang.org/std/convert/trait.From.html):
> One should always prefer implementing From over [Into](https://doc.rust-lang.org/std/convert/trait.Into.html) because implementing From automatically provides one with an implementation of [Into](https://doc.rust-lang.org/std/convert/trait.Into.html) thanks to the blanket implementation in the standard library.